keras/utils/multi_gpu_utils.py

```diff
@@ -231,10 +233,20 @@ def get_slice(data, i, parts):
                 for o in range(len(outputs)):
                     all_outputs[o].append(outputs[o])
 
+    # Deduplicate output names to handle Siamese networks.
+    occurrences = Counter(model.output_names)
+    conflict_counter = {n: 0 for n, count in occurrences.items() if count > 1}
+    output_names = []
+    for n in model.output_names:
+        if n in conflict_counter:
+            conflict_counter[n] += 1
+            n += '_%d' % conflict_counter[n]
+        output_names.append(n)
```

Comment:
We could do this more simply without relying on the Counter import:

```python
counter = {}
output_names = []
for name in model.output_names:
    if name in counter:
        counter[name] += 1
        name += '_' + str(counter[name])
    else:
        counter[name] = 0
    output_names.append(name)
```

Reply:
sure, I can remove the Counter. For the unit-test, I'll use the snippet provided in the description of the PR and expect it will not throw an exception.

Reply:
@fchollet I just noticed that your snippet does not only remove the Counter import, but also changes the proposed naming convention. Based on the example I provided the `['add_1', 'nested', 'nested']` will become `['add_1', 'nested', 'nested_1']` instead of `['add_1', 'nested_1', 'nested_2']`. What that intentional? If not, I can still lose the Counter but the logic will not simplify.

Reply:
Your comment made me notice that the name unification protocol for unnamed layers was different in multi-backend Keras and tf.keras...

multi-backend Keras: `dense_1`, `dense_2`, `dense_3`
tf.keras: `dense`, `dense_1`, `dense_2`

In this specific case (multi_gpu_model) I think we should go with your proposed naming, so feel free to change the logic.

Reply:
Sounds good.

Just to summarize the plan:
1. I keep my proposed logic on naming to keep it consistent with other parts of Keras.
2. I remove the Counter import and build my own Counter instead. Also refactor stuff where possible to simplify logic.
3. Add the test snippet posted on the description of this PR as a test and require it to throw no exception. I'll also run the test on a Multi-GPU server to ensure it passes (already did but reconfirm on final version of PR).